### ISSUE TITLE
Implement CSS Houdini Worklets

### ISSUE THREAD

**Issue by asajeffrey** [servo/servo#16206]
Draft spec: https://drafts.css-houdini.org/worklets/
Houdini CSS issue for servo: #13557 

cc @tschneidereit @metajack @jdm @SimonSapin

**Comment by asajeffrey** [comment]
This came up in the context of discussions about the embedding API, as a standards-aligned way to provide strong consistency guarantees between embedding JS code, layout and webrender.

**Comment by fitzgen** [comment]
.

**Comment by tschneidereit** [comment]
cc @Gozala @paulrouget

**Comment by asajeffrey** [comment]
On irc, @tschneidereit pointed out the code idempotence requirement: https://drafts.css-houdini.org/worklets/#code-idempotency, which is not actually statelessness, but is very close. In particular, a UA is allowed to generate a fresh global for each execution. Not sure what this would do for efficiency though, as scripts might assume that fresh globals aren't allocated very often, e.g. to perform some script initialization.

**Comment by metajack** [comment]
Fresh globals are reportedly too expensive to actually do this, so there is language about modified designs (ie, rotating two pre-allocated globals) added. If the implementation does nothing to preserve idempotency, we are likely to lose it, and any potential for parallelism along with it.

**Comment by asajeffrey** [comment]
My current thinking of how to implement this is to provide a per-process thread pool for executing worklets. The main challenge would be to try to avoid GC during worklet execution, which we could do by monitoring their memory use, and removing high-memory-usage executors from the pool while they gc.

**Comment by asajeffrey** [comment]
@metajack yes, generating a fresh global on every execution might be too expensive, but possibly we could freshen the global periodically, e.g. when the executor is GC'ing?

**Comment by asajeffrey** [comment]
IRC chat with @tschneidereit and @metajack: http://logs.glob.uno/?c=mozilla%23servo&s=31+Mar+2017&e=31+Mar+2017#c643173

TL;DR we'd like to get to an API where there's an explicit initialization phase, during which script has r/w access to the realm, followed by execution, which is in a frozen realm (https://github.com/FUDCo/proposal-frozen-realms). But we can't block worklets on frozen realms, so we're left with just spec language, which doesn't help if webcompat ends up just being everyone assumes stateful worklets.

**Comment by asajeffrey** [comment]
IRC chat with @jdm: http://logs.glob.uno/?c=mozilla%23servo&s=3+Apr+2017&e=3+Apr+2017#c644051

TL;DR: worklets are based on ES6 modules, which aren't in SM yet.

**Comment by asajeffrey** [comment]
IRC chat with @SimonSapin: http://logs.glob.uno/?c=mozilla%23servo&s=5+Apr+2017&e=5+Apr+2017#c645966

TL;DR: stylo shouldn't be a big factor.
